File: services/rust/poem/src/routes/middleware/chaos.rs
use crate::routes::errors::AppError;
use poem::{Endpoint, IntoResponse, Request, Response, Result, error::ResponseError};
use rand::{RngExt, rng};
use std::time::Duration;
use tokio::time::sleep;
use tracing::instrument;

#[instrument(skip_all)]
pub async fn chaos_middleware<E: Endpoint>(next: E, req: Request) -> Result<Response> {
    if rng().random_range(0..=100) < 5 {
        let duration = Duration::from_micros(rng().random_range(500..=1500));
        sleep(duration).await;
    }
    if rng().random_range(0..=100) < 5 {
        return Ok(AppError::Internal.as_response());
    }

    next.call(req).await.map(IntoResponse::into_response)
}

File: services/go/gin/routes/middleware/chaos.go
package middleware

import (
	"gin/routes"
	"math/rand/v2"
	"time"

	"github.com/gin-gonic/gin"
)

func ChaosMiddleware() gin.HandlerFunc {
	return func(c *gin.Context) {
		if rand.IntN(101) < 5 {
			delay := time.Duration(500+rand.IntN(1001)) * time.Microsecond
			time.Sleep(delay)
		}
		if rand.IntN(101) < 5 {
			routes.AppErrors.Internal.Error(c)
			c.Abort()
			return
		}

		c.Next()
	}
}
